lms/djangoapps/email_marketing/tests/test_signals.py

```diff
@@ -0,0 +1,169 @@
+        update_user.delay(self.user.username)
+        self.assertTrue(mock_log_error.called)
+
+    @patch('email_marketing.tasks.SailthruClient.api_post')
+    @patch('email_marketing.tasks.User.objects.get')
+    def test_change_email(self, mock_user_get, mock_sailthru):
+        """
+        test async method in tasks that changes email in Sailthru
+        """
+        mock_user_get.return_value = self.user
+        mock_sailthru.return_value = SailthruResponse(JsonResponse({'ok': True}))
+        self.user.email = "newemail@test.com"
+        update_user_email.delay(self.user.username, "test@edx.org")
+        self.assertEquals(mock_sailthru.call_args[0][0], "user")
+        userparms = mock_sailthru.call_args[0][1]
+        self.assertEquals(userparms['key'], "email")
+        self.assertEquals(userparms['id'], "test@edx.org")
+        self.assertEquals(userparms['keys']['email'], "newemail@test.com")
+
+    @patch('email_marketing.tasks.log.error')
+    @patch('email_marketing.tasks.SailthruClient.api_post')
+    @patch('email_marketing.tasks.User.objects.get')
+    def test_error_logging1(self, mock_user_get, mock_sailthru, mock_log_error):
+        """
+        Ensure that error returned from Sailthru api is logged
+        """
+        mock_user_get.return_value = self.user
+        mock_sailthru.return_value = SailthruResponse(JsonResponse({'error': 100, 'errormsg': 'Got an error'}))
+        update_user_email.delay(self.user.username, "newemail2@test.com")
+        self.assertTrue(mock_log_error.called)
+
+    @patch('lms.djangoapps.email_marketing.tasks.update_user.delay')
+    def test_modify_field1(self, mock_update_user):
+        """
+        try updating user field
+        """
+        email_marketing_register_user(None, user=self.user, profile=self.profile)
+        self.assertTrue(mock_update_user.called)
+
+    @patch('lms.djangoapps.email_marketing.tasks.update_user.delay')
+    def test_modify_field2(self, mock_update_user):
+        """
+        try updating user field
+        """
+        email_marketing_user_field_changed(None, self.user, table='auth_userprofile', setting='gender', new_value='f')
+        self.assertTrue(mock_update_user.called)
+
+    @patch('lms.djangoapps.email_marketing.tasks.update_user.delay')
+    def test_modify_field3(self, mock_update_user):
+        """
+        try updating profile field
+        """
+        email_marketing_user_field_changed(None, self.user, table='auth_user', setting='is_active', new_value=1)
+        self.assertTrue(mock_update_user.called)
+
+    @patch('lms.djangoapps.email_marketing.tasks.update_user.delay')
+    def test_modify_field4(self, mock_update_user):
+        """
+        try updating unsupported field
+        """
+        email_marketing_user_field_changed(None, self.user, table='auth_userprofile', setting='shoe_size', new_value=1)
+        self.assertFalse(mock_update_user.called)
```

Comment:
A pattern we use elsewhere in the codebase is to use the `ddt` library (https://github.com/txels/ddt). This would let you specify a list like

    @ddt.data(('auth_user_profile', 'gender', 'f', True), ('auth_user', 'is_active', 1, True), ('auth_userprofile', 'shoe_size', 1, False))
    @ddt.unpack
    def test_modify_field(self, mock_update_user, table, setting, value, result):
        email_marketing_user_field_changed(None, self.user, table=table, setting=setting, new_value=value) 
        self.assertEqual(mock_update_user.called, result)